test/test_torch.py

```diff
@@ -98,6 +99,47 @@ def __exit__(self, *args):
         pass
 
 
+def _rebuild_subclass(type_, data, requires_grad, backward_hooks):
+    param = type_(data, requires_grad)
+    # NB: This line exists only for backwards compatibility; the
+    # general expectation is that backward_hooks is an empty
+    # OrderedDict.  See Note [Don't serialize hooks]
+    param._backward_hooks = backward_hooks
+
+    return param
+
+
+class TensorSubclass(torch.Tensor):
+
+    def __new__(cls, data=None, requires_grad=False):
+        if data is None:
+            data = torch.Tensor()
+        self = torch.Tensor._make_subclass(cls, data, requires_grad)
+        return self
```

Comment:
Do you have an explanation for how, in general, you should go about defining subclasses of tensor? I see here that you had to explicitly define `__new__`; this is a bit magical, so it would be really helpful to know exactly why this is needed, and how the general pattern works.

Reply:
`TensorSubclass` in `test_torch.py` is identical to Parameter in `torch.nn.parameter.py`. I know you have to define `__deepcopy__`, `__reduce_ex__`.  I think `__reduce_ex__` is doing something for legacy purposes, but I have included it anyways. These are tested in the unit tests.

> "you had to explicitly define __new__; this is a bit magical"

`__new__(cls, ...)` takes `cls` (a class object) as the first argument and defines the constructor of a **class**. You pass the `PyTypeObject* cls` to `THPVariable_make_subclass()` which defines the Python (ptype) of the class.

Example:
```python
my_class = MyTensorSubClass  # A Class (or type)
```

`__init__(self, )` takes `self` ( an instance of the class) as the first argument and defines a constructor of an **object**.

Example:
```python
my_obj = MyTensorSubClass()  # An instance of the Class.
```

The terminology is really weird here because in Python an object, class, function, method, ..., etc are all objects.